design-proposals/sdn-ipam-secondary-nets.md

```diff
@@ -0,0 +1,450 @@
+
+KubeVirt -->> user: OK
+
+note over OVN-K: notices a pod was deleted
+OVN-K ->> OVN-K: ips = teardownOVNEntities()
+```
+
+OVN-Kubernetes will **not** delete the `IPAMClaim` CRs for the deleted VM pod.
+The IP addresses allocated to the VM will **not** be released to the IP pool
+when there's a corresponding `IPAMClaim` for the attachment being deleted.
+
+#### Starting a (previously stopped) Virtual Machine
+This flow is - from a CNI perspective - quite similar to the
+[Creating a VM flow](#creating-a-virtual-machine):
+1. the workload controller (KubeVirt) templates the pod, featuring the required
+   `IPAMClaim` references in each network selection element.
+2. OVN-Kubernetes will read the network selection element; if it sees an
+   `IPAMClaim` reference, it will attempt to read it from the datastore.
+3. if the read `IPAMClaim` features IP addresses in its status, OVN-Kubernetes
+   will configure the interface with those IPs. If the status doesn't feature
+   IP addresses, the CNI plugin will retry later, with exponential backoff.
+
+```mermaid
+sequenceDiagram
+
+  actor user
+  participant KubeVirt
+  participant Kubernetes
+  participant CNI
+
+  user->>KubeVirt: startVM(vmName)
+
+  note over KubeVirt: podName := "launcher-<vm name>"
+  KubeVirt->>Kubernetes: createPod(name=podName, ipam-claims=[claimName])
+
+  Kubernetes->>CNI: CNI ADD
+  note over CNI: this CNIs has an IPAMClaims informer
+
+  CNI->>CNI: ipamClaim := readIPAMClaim(claimName)
+  CNI->>CNI: configIface(ipamClaim.Status.IPs)
+
+  CNI-->>Kubernetes: OK
+
+  Kubernetes-->>KubeVirt: OK
+
+
+  KubeVirt-->>user: OK
+```
+
+When creating the pod, the OVN-Kubernetes IPAM module finds existing
+`IPAMClaim`s for the workload. It will thus use those already reserved
+allocations, instead of generating brand new allocations for the pod where the
+encapsulating object will run. The migration scenario is similar.
+
+#### Removing a Virtual Machine
+This flow is - from a CNI perspective - quite similar to the
+[Stopping a VM flow](#stopping-a-virtual-machine). The main difference is after
+the VM is deleted, Kubernetes Garbage Collection will kick in, and remove the
+orphaned `IPAMClaim`s. OVN-Kubernetes will need to react to the `IPAMClaim`
+delete event, to return those IP addresses to the respective IP pools.
+
+```mermaid
+sequenceDiagram
+
+  actor user
+  participant KubeVirt
+  participant Kubernetes
+  participant CNI
+  participant IPAM CNI
+
+  user->>KubeVirt: deleteVM(vmName)
+
+  KubeVirt->>Kubernetes: deletePod(launcher-vmName)
+
+  Kubernetes->>CNI: CNI DEL
+  CNI-->>Kubernetes: OK
+
+  Kubernetes-->>KubeVirt: OK
+
+  KubeVirt-->>user: OK
+
+  loop for every orphaned IPAMClaim
+    Kubernetes->>Kubernetes: deleteIPAMClaim()
+  end
+
+  note over IPAM CNI: this CNI has an IPAMClaim informer
+
+Kubernetes->>IPAM CNI: removeEvent(ipamClaim)
+IPAM CNI->>IPAM CNI: returnToPool(ipamClaim.Status.IPs)
+```
+
+#### Hot-plug a VM interface
+This flow is exactly the same as
```

Comment:
Please help me understand your expectation. I don't understand what you mean with integration point.

Reply:
The location in the VMI controller you intend to process the logic.
Or any other place you need to place logic at to reconcile this.

E.g. if you need to integrate this at the VM controller and not at the VMI controller, it matters.

Reply:
I'm pretty sure the plan is to integrate at the VMI controller, given we also want to support IP stickiness for VMI migration.

Reply:
I am not in favor of supporting it at the VMI level, but it will be interesting to understand the exact need and usefulness. Te migration is covered if initiated from a VM object.

We have several features, like hotplug, that are supported only at the VM level.
I'm not seeing the added value of supporting VMI independently, but if in practice we do have such users whith common usage, I should reconsider my position on this.

Reply:
(must-have) Support sticky IPs on VMs
(should-have) Support sticky IPs on VMIs

So not a requirement per-se.

The reason why I want sticky IPs on VMIs is to have a clear support story "sticky IPs are supported" without the condition "unless you are using VMIs". If this is simple to achieve (which from what I understood from our offline discussion it is), then I think it's worth having this feature available everywhere.

Alternatively, the support story could be "only VMs are supported, don't use VMIs, ever". In that case, I would be happy not having sticky IPs on VMIs.

Ultimately it's @AlonaKaplan's call - do we treat VMIs as a an API that's meant to be used by users, or no.

Reply:
[About](https://github.com/kubevirt/community/pull/279#discussion_r1537653418) if to use the logic on VM or VMI controller 
Few reasons why imho we should use VMI controller

1. Only VMI controller knows when the pod is [created](https://github.com/kubevirt/kubevirt/blob/055c6e0491fa93befa6372ca4d367916cabcb5af/pkg/virt-controller/watch/vmi.go#L1196).
Creating the claims there, just before the pod is created, removes the need to recreate / re-validate on every VM reconcile iteration, and it is much better design imho.

2. VMI is the one that already reads the NAD, it is more natural for it to read the NAD, and i dont think we should add NAD reading also to VM.

3. For hot (un)plug we would need also to sync the claims, the annotations updating is the point that already calculates the difference that need to be done (and it is on VMI controller).

Reply:
@AlonaKaplan the VMI only flows are pretty much the same. 

The notable exceptions are:
- hotplug is not supported (no hotplug on VMI)
- restart is not supported (restart essentially spuns up a new VMI)

Everything else is pretty much the same - with the notable exception the IPAMClaim's reference **owner** is a VMI, instead of a VM.

@oshoval please add more detail if you think it is important

Reply:
you covered it all
supporting VMI standalone is even easier than not supporting it, because we don't need to add a check if there is VM owner or not when updating the annotation / creating claims.
The only condition that exists about those is if there is no VM owner, then set the ownerRef as VMI as you said.